File: backend/src/deaths_door/routes/timer.py
from fastapi import APIRouter, HTTPException, Path, Query
from pydantic import BaseModel, Field

from ..config import Config
from ..timer_state import TimerState


class TimerStateResponse(BaseModel):
    """Current state of the countdown timer."""

    is_running: bool = Field(..., description="Whether the timer is currently running")
    seconds: int = Field(..., description="Remaining seconds on the timer", examples=[300])


class TimerOperationResponse(BaseModel):
    """Response after a timer operation."""

    status: str = Field(..., description="Operation status", examples=["success"])
    is_running: bool = Field(..., description="Whether the timer is currently running")
    seconds: int = Field(..., description="Current seconds on the timer", examples=[300])


def validate_timer_seconds(seconds: int, allow_negative: bool = False) -> None:
    """Validate timer seconds are within acceptable range."""
    min_seconds = -Config.TIMER_MAX_SECONDS if allow_negative else 0
    max_seconds = Config.TIMER_MAX_SECONDS

    if not (min_seconds <= seconds <= max_seconds):
        range_desc = f"between {min_seconds} and {max_seconds}"
        raise HTTPException(
            status_code=400, detail=f"Timer seconds must be {range_desc}"
        )


timer = APIRouter(prefix="/timer", tags=["Timer"])
state = TimerState()


@timer.get("/fetch")
async def fetch_timer() -> TimerStateResponse:
    """Fetch the current state of the timer."""
    return TimerStateResponse(
        is_running=await state.get_is_running(),
        seconds=await state.get_seconds(),
    )


@timer.get("/set/{seconds}", responses={400: {"description": "Invalid timer seconds"}})
async def set_timer(
    seconds: int = Path(
        ...,
        description="Number of seconds to set on the timer",
        examples=[300],
    )
) -> TimerOperationResponse:
    """Set the timer to the given number of seconds and start it."""
    validate_timer_seconds(seconds)
    await state.set_seconds(seconds)
    await state.set_is_running(True)
    return TimerOperationResponse(
        status="success", is_running=True, seconds=seconds
    )


@timer.get("/add/{seconds}", responses={400: {"description": "Invalid timer seconds"}})
async def add_seconds(
    seconds: int = Path(
        ...,
        description="Number of seconds to add (can be negative)",
        examples=[60, -30],
    )
) -> TimerOperationResponse:
    """Add the given number of seconds to the timer (can be negative to subtract)."""
    validate_timer_seconds(seconds, allow_negative=True)
    await state.add_seconds(seconds)
    return TimerOperationResponse(
        status="success",
        is_running=await state.get_is_running(),
        seconds=await state.get_seconds(),
    )


@timer.get("/start", responses={400: {"description": "Invalid timer seconds"}})
async def start_timer(
    seconds: int | None = Query(
        None,
        description="Optional number of seconds to set before starting the timer",
        examples=[300, 60],
    )
) -> TimerOperationResponse:
    """Start the timer, optionally setting seconds first."""
    if seconds is not None:
        validate_timer_seconds(seconds)
        await state.set_seconds(seconds)
    await state.set_is_running(True)
    return TimerOperationResponse(
        status="success",
        is_running=True,
        seconds=await state.get_seconds(),
    )


@timer.get("/stop")
async def stop_timer() -> TimerOperationResponse:
    """Stop the timer."""
    await state.set_is_running(False)
    return TimerOperationResponse(
        status="success",
        is_running=False,
        seconds=await state.get_seconds(),
    )

File: backend-rust/src/routes/timer.rs
//! Routes for the countdown timer and Live Activity push tokens.

use axum::extract::{Path, Query, State};
use axum::Json;
use serde::{Deserialize, Serialize};
use serde_json::json;
use utoipa::ToSchema;
use utoipa_axum::router::OpenApiRouter;
use utoipa_axum::routes;

use crate::app::{AppError, AppJson, AppResult, AppState};
use crate::config;

pub fn router() -> OpenApiRouter<AppState> {
    OpenApiRouter::new()
        .routes(routes!(fetch_timer))
        .routes(routes!(set_timer))
        .routes(routes!(add_seconds))
        .routes(routes!(start_timer))
        .routes(routes!(stop_timer))
        .routes(routes!(register_push_token))
}

#[derive(Serialize, ToSchema)]
pub struct TimerStateResponse {
    pub is_running: bool,
    pub seconds: i64,
}

#[derive(Serialize, ToSchema)]
pub struct TimerOperationResponse {
    pub status: String,
    pub is_running: bool,
    pub seconds: i64,
}

fn validate_timer_seconds(seconds: i64, allow_negative: bool) -> AppResult<()> {
    let min = if allow_negative {
        -config::TIMER_MAX_SECONDS
    } else {
        0
    };
    let max = config::TIMER_MAX_SECONDS;
    if seconds < min || seconds > max {
        return Err(AppError::bad_request(format!(
            "Timer seconds must be between {min} and {max}"
        )));
    }
    Ok(())
}

/// Fetch the current timer state.
#[utoipa::path(
    get, path = "/timer/fetch", tag = "Timer",
    responses((status = 200, description = "Current timer state", body = TimerStateResponse))
)]
async fn fetch_timer(State(state): State<AppState>) -> Json<TimerStateResponse> {
    Json(TimerStateResponse {
        is_running: state.timer.get_is_running().await,
        seconds: state.timer.get_seconds().await,
    })
}

/// Set the timer to a specific number of seconds and start it.
#[utoipa::path(
    get, path = "/timer/set/{seconds}", tag = "Timer",
    params(("seconds" = i64, Path, description = "Seconds to set, 0..=3600")),
    responses(
        (status = 200, description = "Timer set", body = TimerOperationResponse),
        (status = 400, description = "out of range")
    )
)]
async fn set_timer(
    State(state): State<AppState>,
    Path(seconds): Path<i64>,
) -> AppResult<Json<TimerOperationResponse>> {
    validate_timer_seconds(seconds, false)?;
    state.timer.set_seconds(seconds).await;
    state.timer.set_is_running(true).await;
    Ok(Json(TimerOperationResponse {
        status: "success".to_string(),
        is_running: true,
        seconds,
    }))
}

/// Add (or subtract) seconds from the timer.
#[utoipa::path(
    get, path = "/timer/add/{seconds}", tag = "Timer",
    params(("seconds" = i64, Path, description = "Seconds to add (may be negative)")),
    responses(
        (status = 200, description = "Timer adjusted", body = TimerOperationResponse),
        (status = 400, description = "out of range")
    )
)]
async fn add_seconds(
    State(state): State<AppState>,
    Path(seconds): Path<i64>,
) -> AppResult<Json<TimerOperationResponse>> {
    validate_timer_seconds(seconds, true)?;
    state.timer.add_seconds(seconds).await;
    Ok(Json(TimerOperationResponse {
        status: "success".to_string(),
        is_running: state.timer.get_is_running().await,
        seconds: state.timer.get_seconds().await,
    }))
}

#[derive(Deserialize, utoipa::IntoParams)]
pub struct StartQuery {
    pub seconds: Option<i64>,
}

/// Start the timer, optionally setting it to a number of seconds first.
#[utoipa::path(
    get, path = "/timer/start", tag = "Timer",
    params(StartQuery),
    responses(
        (status = 200, description = "Timer started", body = TimerOperationResponse),
        (status = 400, description = "out of range")
    )
)]
async fn start_timer(
    State(state): State<AppState>,
    Query(query): Query<StartQuery>,
) -> AppResult<Json<TimerOperationResponse>> {
    if let Some(seconds) = query.seconds {
        validate_timer_seconds(seconds, false)?;
        state.timer.set_seconds(seconds).await;
    }
    state.timer.set_is_running(true).await;
    Ok(Json(TimerOperationResponse {
        status: "success".to_string(),
        is_running: true,
        seconds: state.timer.get_seconds().await,
    }))
}

/// Stop the timer.
#[utoipa::path(
    get, path = "/timer/stop", tag = "Timer",
    responses((status = 200, description = "Timer stopped", body = TimerOperationResponse))
)]
async fn stop_timer(State(state): State<AppState>) -> Json<TimerOperationResponse> {
    state.timer.set_is_running(false).await;
    Json(TimerOperationResponse {
        status: "success".to_string(),
        is_running: false,
        seconds: state.timer.get_seconds().await,
    })
}

#[derive(Deserialize, ToSchema)]
pub struct PushTokenRequest {
    pub push_token: String,
}

/// Register a Live Activity push token.
#[utoipa::path(
    post, path = "/timer/push_token", tag = "Timer",
    request_body = PushTokenRequest,
    responses((status = 200, description = "Token registered"))
)]
async fn register_push_token(
    State(state): State<AppState>,
    AppJson(req): AppJson<PushTokenRequest>,
) -> Json<serde_json::Value> {
    state.timer.apns().register_token(req.push_token);
    Json(json!({ "status": "registered" }))
}
